x/pairing/keeper/scores/geo_req.go

```diff
@@ -0,0 +1,216 @@
+package scores
+
+import (
+	"fmt"
+	"sort"
+
+	commontypes "github.com/lavanet/lava/common/types"
+	"github.com/lavanet/lava/utils"
+	epochstoragetypes "github.com/lavanet/lava/x/epochstorage/types"
+	"github.com/lavanet/lava/x/pairing/types"
+	planstypes "github.com/lavanet/lava/x/plans/types"
+)
+
+// geo requirement that implements the ScoreReq interface
+type GeoReq struct {
+	Geo uint64
+}
+
+const (
+	geoReqName    = "geo-req"
+	maxGeoLatency = 10000 // highest geo cost < 300
+	minGeoLatency = 1
+)
+
+// Score calculates the geo score of a provider (using the GEO_LATENCY_MAP)
+// The score is (maxGeoLatency / minLatency)^geoWeight
+// Note: the geo slots are created so that the GeoReq has a single bit geolocation. This function assumes that this is the case
+func (gr GeoReq) Score(provider epochstoragetypes.StakeEntry) uint64 {
+	if !types.IsGeoEnumSingleBit(int32(gr.Geo)) {
+		utils.LavaFormatError("provider geo req is not single bit", fmt.Errorf("invalid geo req"),
+			utils.Attribute{Key: "geoReq", Value: gr.Geo},
+		)
+		return calculateCostFromLatency(maxGeoLatency)
+	}
+
+	// check if the provider supports the required geolocation
+	if gr.Geo&^provider.Geolocation == 0 {
+		return calculateCostFromLatency(minGeoLatency)
+	}
+
+	providerGeoEnums := types.GetGeolocationsFromUint(int32(provider.Geolocation))
+	_, cost := GetGeoCost(planstypes.Geolocation(gr.Geo), providerGeoEnums)
+
+	return cost
+}
+
+func (gr GeoReq) GetName() string {
+	return geoReqName
+}
+
+// Equal() used to compare slots to determine slot groups
+func (gr GeoReq) Equal(other ScoreReq) bool {
+	otherGeoReq, ok := other.(GeoReq)
+	if !ok {
+		return false
+	}
+
+	return otherGeoReq.Geo == gr.Geo
+}
+
+func GetGeoReqsForSlots(policy planstypes.Policy) []GeoReq {
+	geoReqsForSlots := []GeoReq{}
+
+	policyGeoEnums := types.GetGeolocationsFromUint(int32(policy.GeolocationProfile))
+	switch {
+	// TODO: implement the case below
+	// case len(policyGeoEnums) > int(policy.MaxProvidersToPair):
+	default:
+		for i := 0; i < int(policy.MaxProvidersToPair); i++ {
+			geoReq := GeoReq{Geo: uint64(policyGeoEnums[i%len(policyGeoEnums)])}
+			geoReqsForSlots = append(geoReqsForSlots, geoReq)
+		}
+	}
+
+	return geoReqsForSlots
+}
+
+// a single geolocation and the latency to it (in millieseconds)
+type GeoLatency struct {
+	geo     planstypes.Geolocation
+	latency uint64
+}
+
+func (gl GeoLatency) Less(other GeoLatency) bool {
+	return gl.latency < other.latency
+}
+
+// GetGeoCost() finds the minimal latency between the required geo and the provider's supported geolocations
+func GetGeoCost(reqGeo planstypes.Geolocation, providerGeos []planstypes.Geolocation) (minLatencyGeo planstypes.Geolocation, minLatencyCost uint64) {
+	geoLatencies := []GeoLatency{}
+	latencies := []uint64{}
+	for _, pGeo := range providerGeos {
+		geoLatency := getGeoLatency(reqGeo, pGeo)
```

Comment:
change GEO_LATENCY_MAP to be a map of maps, and then all you'll need here would be something like:
```
    var minGeo GeoLatency
    minLatency := math.MaxUint64
    for _, pGeo := range providerGeos {
        if inner, ok := GEO_LATENCY_MAP[repGeo]; ok {
            if latency, ok := inner[pGeo]; ok {
                if latency < minLatency {
                    minGeo := pGeo
                    minLatency = latency
                }
            }
        }
    }
    if minLatency == math.MaxUint64 {
        ... // no geolatencies found
    }
    ... // min geo latency found
```

Reply:
done. see https://github.com/lavanet/lava/commit/bd878b42cfcb830adda0109d6160f15b7d635abf